[[ FFI ]] Obj-C

runrevmark opened the pull request:
This patch contains the basic support for Obj-C FFI in LCB.

It provides an ObjcObject type (non-foreign) and ObjcRetainedId, ObjcId and ObjcAutoreleasedId foreign types to help manage lifetimes of objects.

Foreign handlers can now use 'objc' as the language, and in that case can bind to both instance and class methods of Objective-C objects.

There is also basic support for 'proxy' objects, allowing LCB to use the Target-Action pattern used ubiquitously in Cocoa and CocoaTouch.

The current implementation has been used to implement a basic native Mac button - which will send a mouseUp message when clicked, and allows the label to be set. The button also hooks into enabled and font host properties.

livecodeali commented on extensions/widgets/macbutton/macbutton.lcb:
```diff
@@ -0,0 +1,214 @@
+Value (string): The string to use as the button label
+
+Example:
+    set the label of widget "Android Button" to "Click me!"
```
Mac Button

livecodeali commented on extensions/widgets/macbutton/macbutton.lcb:
```diff
@@ -0,0 +1,214 @@
+	variable tButtonView as ObjcObject
+	put ObjC_NSButtonAlloc() into tButtonView
+	put ObjC_NSButtonInit(tButtonView) into tButtonView
+	ObjC_NSButtonSetButtonType(tButtonView, 7)
```
You could maybe put all these button types in as constants and then use the named one, or just say that this 7 comes from https://developer.apple.com/documentation/appkit/nsbutton.buttontype/2097066-momentarypushin

livecodeali commented:
@livecode-vulcan review ok 943e4ca

livecode-vulcan commented:
💙 review by @livecodeali ok 943e4ca85c11a9597ff10caf4a1fa245f7c2f278

livecode-vulcan commented:
😞 test failure 943e4ca85c11a9597ff10caf4a1fa245f7c2f278

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/477)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/492)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/464)
* try-community-universal-ios-iphoneos10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/407)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/410)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/472)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/478)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/475)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/490)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/475)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/485)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/503)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/471)

runrevmark committed:
[[ Bug 20058 ]] Ensure changing name of widget sends notification

This patch ensures that 'recompute()' is called if the name of a
widget changes. This results in 'OnParentPropertyChanged' being
called in the widget, allowing it to update its state.

runrevmark committed:
[[ Bug 20059 ]] Ensure native layer geometry is updated before painted

This patch ensures that if geometry has been deferred (due to the layer
being hidden) then it is applied before rendering to context. This means
that layers look correct when first created in Edit mode.

runrevmark committed:
[[ LCB ]] Add access to font handle of CanvasFont

This patch adds MCCanvasFontGetHandle() which allows the internal
font handle to be fetched from LCB. For example, on Mac, the handle
is an NSFont*.

runrevmark committed:
[[ Foundation ]] Remove 'bool' primitive type

This patch removes the bool primitive type as it doesn't exist - 'bool'
is one of the primitive native int types in all calling conventions.

runrevmark committed:
[[ Bug 20063 ]] Render Mac native layer snapshot correctly

This patch ensures that the snapshot generated for rendering
native layers in edit mode use the size of the view in points,
rather than the size of the snapshot in (device) pixels.

runrevmark committed:
[[ FFI ]] Add Obj-C FFI

This patch adds a collection of types to use for binding to Obj-C methods
as well as the binding string support for hooking up to Obj-C object instance
and class methods.

runrevmark committed:
[[ Bug 19945 ]] Correct Mac native layer snapshots

This patch corrects an off-by-one-pixel error in the rendering
of Mac native layer snapshots.

runrevmark committed:
[[ LCB ]] Add a basic native Mac button widget

This patch uses the obj-c FFI support to add a native Mac button
widget.

The widget can currently have its label set, and will send
`mouseUp` when it is clicked.

runrevmark committed:
[[ LCB ]] Fix incorrect names of UIntPtr and SIntPtr in mapping

This patch corrects the incorrect names used when building 'builtin'
modules for the mapping of the UIntPtr and SIntPtr types.

runrevmark commented:
This latest updated added a missing ';' in script-execute.cpp which was only affecting iOS.

@livecode-vulcan review ok 44eded1

livecode-vulcan commented:
💙 review by @runrevmark ok 44eded14f79d3918e438c690476db8774239467c

livecode-vulcan commented:
😎 test success 44eded14f79d3918e438c690476db8774239467c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/480)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/495)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/467)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/410)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/413)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/475)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/481)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/478)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/493)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/478)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/488)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/506)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/474)
